doc/error_handling.md

```diff
@@ -0,0 +1,285 @@
+If the error cannot be handled at a higher level by the caller, the error handler needs to be used.
+
+Examples include wrapping third party API functions that return error codes or obtaining a value from a container when this can fail for some reason (e.g. container is empty). If no additional information about the error is available or required, **cxx::optional<T>** can be used instead.
+
+It may be worth to consider renaming cxx::expected to cxx::result in the future, which is more in line with languages such as *Rust* and conveys the meaning more clearly.
+
+## Error Handling in posh
+Error logging shall be done by the logger only, no calls to std::cerr or similar should be performed.
+
+All the methods presented (cxx::expected, Expects and Ensures and the error handler) can be used in posh. The appropriate way depends on the type of error scenario (cf. the respective sections for examples). The error handler should be considered the last option.
+
+## Error Handling in utils
+Error logging is currently done by calls to cerr. In the future those might be redirected to the logger.
+
+The error handler cannot be used in utils. 
+
+Whether it is appropriate to use std::expected even if STL compatility is broken by doing so depends on the circumstances and needs to be discussed on a case-by-case basis. If the function has no STL counterpart std::expected can be used freely to communicate potential failure to the caller.
+
+It should be noted that since currently Expects and Ensures are active at release mode, prolific usage of these will incur a runtime cost. Since this is likely to change in the future, it is still advised to use them to document the developers intentions.
+
+## Interface for 3rd Party Code
+
+Error handler as well as logger shall be able to use or redirect to 3rd party error handling or logging libraries in the future. Currently this is not supported.
+
+# Usage
+
+## Logger
+The logger can be used similar to the streams in the C++ standard API.
+To select the log level, the corresponding logger has to be used, e.g. LogErr, LogWarn etc.
+
+```
+LogWarn() << "log message " << someValue << "log message continued";
+```
+
+## Error Handler
+The most general use case is the following
+```
+if(noError()) {
+    //handle the regular case
+} else {
+    auto callback = []() { //some error handling callback};
+    errorHandler(Error::kSOME_ERROR_CODE, callback, ErrorLevel::SEVERE);
+}
+```
+
+If no callback or error level are specified, the error level is assumed to be FATAL by default.
+```
+errorHandler(Error::kSOME_ERROR_CODE)
+```
+
+If no callback but an error level is desired, a nullptr has to be provided for the callback.
+```
+errorHandler(Error::kSOME_ERROR_CODE, nullptr, ErrorLevel::MODERATE);
+```
+We should consider changing the order of arguments in a future design (callback and additional arguments last). This can be done in a redesign were we provide additional information about the error location as well.
```

Comment:
Note that this would require using a macro for this. I also have some ideas for a very general highly customizable/extendable error/assert handling which, when configured properly, can be optimized away completely or on a level by level basis. This requires further investigation and is future work.

Reply:
Move this and similar points to a Future improvements section.